File: matlab/+spine/+systems/system_dispersive_readout.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Author: J.P.G. van Dijk                                                 %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

classdef system_dispersive_readout < handle
    
    properties
        
        % Boltzmann constant (eV)
        kB = 8.62e-5;
        
        % Hamiltonian Properties
        electron_temperature = 0;
        tunnel_rate = 0;
        lever_arm = 0;
        
        % Hamiltonian Control
        gate_voltage = 0;
        
        % Plot handles
        p = gobjects(1, 1);
        
    end
    
    methods
        
        function obj = system_dispersive_readout()
            
        end
        
        function dimension = getDimension(obj)
            dimension = 2;
        end
        
        function state = initialize(~, P)
            state = [1; P];
        end
        
        function [P] = measure(~, state)
            P = state(2);
        end
        
        function H = updateHamiltonian(obj)
            H = 1i * [0, 0; obj.tunnel_rate / (1 + exp(obj.lever_arm * obj.gate_voltage/obj.kB/obj.electron_temperature)), -obj.tunnel_rate];
        end
        
        % Hamiltonian Property Getters/Setters
        function setElectronTemperature(obj, electron_temperature)
            obj.electron_temperature = electron_temperature;
        end
        
        function electron_temperature = getElectronTemperature(obj)
            electron_temperature = obj.electron_temperature;
        end
        
        function setTunnelRate(obj, tunnel_rate)
            obj.tunnel_rate = tunnel_rate;
        end
        
        function tunnel_rate = getTunnelRate(obj)
            tunnel_rate = obj.tunnel_rate;
        end
        
        function setLeverArm(obj, lever_arm)
            obj.lever_arm = lever_arm;
        end
        
        function lever_arm = getLeverArm(obj)
            lever_arm = obj.lever_arm;
        end
        
        % Hamiltonian Control Setters
        function setGateVoltage(obj, gate_voltage)
            obj.gate_voltage = gate_voltage;
        end
        
        % Plot
        function plot(obj, state_or_U, t)

            % Init, rotate, measure
             if (size(state_or_U, 2) == 1)
                state = state_or_U;
            else
                state = state_or_U * obj.initialize(0);
             end
            [P] = obj.measure(state);

            % Plot
            if (isempty(fieldnames(get(obj.p))))
                figure();
                obj.p(1) = plot(t, P);
            else
                obj.p(1).XData = [obj.p(1).XData, t];
                obj.p(1).YData = [obj.p(1).YData, P];
            end
    
        end
        
    end
    
end

File: cpp/spine/systems/system_dispersive_readout.cpp
/***************************************************************
* Author: J.P.G. van Dijk
***************************************************************/

#include "system_dispersive_readout.h"

using namespace spine;

spine::systems::system_dispersive_readout::system_dispersive_readout()
{
	
}

spine::systems::system_dispersive_readout::~system_dispersive_readout()
{
#ifdef PLOT
	delete p;
#endif
}

unsigned int spine::systems::system_dispersive_readout::getDimension()
{
	return 2;
}

void spine::systems::system_dispersive_readout::initialize(complex * state, realnum P)
{
	state[0].real = 1;
	state[0].imag = 0;
	state[1].real = P;
	state[1].imag = 0;
}

realnum spine::systems::system_dispersive_readout::measure(complex * state)
{
	return state[1].real;
}

void spine::systems::system_dispersive_readout::updateHamiltonian(complex * H)
{
	H[0] = { 0, 0 };
	H[1] = { 0, 0 };
	H[2].real = 0;
	H[2].imag = tunnel_rate / (1 + exp(lever_arm * gate_voltage / kB / electron_temperature));
	H[3].real = 0;
	H[3].imag = -tunnel_rate;
}

#ifdef PLOT

void spine::systems::system_dispersive_readout::plotSetup(unsigned int points, double xmin, double xmax)
{
	this->points = points;
	this->xmin = xmin;
	this->xmax = xmax;
}

void spine::systems::system_dispersive_readout::plotAddU(complex * U, realnum t, bool plot_lab)
{
	unsigned int dim = getDimension();
	complex * statetmp = allocComplex(dim);
	complex * state = allocComplex(dim);

	// Update the overall operation
	initialize(state, 0.0);
	initCopy(dim, statetmp, state);
	multiplyVector(dim, U, statetmp, state);
	plotAdd(state, t, plot_lab);

	freeComplex(state);
	freeComplex(statetmp);
}

void spine::systems::system_dispersive_readout::plotAdd(complex * state, realnum t, bool plot_lab)
{
	realnum P;
	P = measure(state);

	if (p == nullptr)
	{
		p = new Plot(L"Probability", points);
		p->add(0, 0, 255, xmin, xmax, 0, 1);
	}
	p->get(0)->add(t, P);
}

void spine::systems::system_dispersive_readout::plot()
{
	if (p != nullptr)
		p->redraw();
}

#endif

void spine::systems::system_dispersive_readout::setElectronTemperature(realnum value)
{
	electron_temperature = value;
}

realnum spine::systems::system_dispersive_readout::getElectronTemperature()
{
	return electron_temperature;
}

void spine::systems::system_dispersive_readout::setTunnelRate(realnum value)
{
	tunnel_rate = value;
}

realnum spine::systems::system_dispersive_readout::getTunnelRate()
{
	return tunnel_rate;
}
void spine::systems::system_dispersive_readout::setLeverArm(realnum value)
{
	lever_arm = value;
}

realnum spine::systems::system_dispersive_readout::getLeverArm()
{
	return lever_arm;
}
void spine::systems::system_dispersive_readout::setGateVoltage(realnum value)
{
	gate_voltage = value;
}
